### PR TITLE
ldr: Update NRR/NRO structs

### PR DESCRIPTION
**Pull request by VolcaEM**
This was based on Switchbrew pages:

https://switchbrew.org/wiki/NRR

https://switchbrew.org/wiki/NRO

### CODE REVIEW

File: src/core/hle/service/ldr/ldr.cpp [hunk 1]
```diff
@@ -39,47 +39,68 @@ constexpr ResultCode ERROR_NOT_INITIALIZED{ErrorModule::Loader, 87};
...
     u32_le nro_size;
     u32_le flags;
-    u32_le text_offset;
-    u32_le text_size;
-    u32_le ro_offset;
-    u32_le ro_size;
-    u32_le rw_offset;
-    u32_le rw_size;
+    // .text, .ro, .data (yuzu previously called it "rw" instead of "data")
+    std::array<SegmentHeader, 0x3> segment_headers;
```
Comment (lioncash) [inline src/core/hle/service/ldr/ldr.cpp:93]:
```suggestion
    std::array<SegmentHeader, 3> segment_headers;
```

There's also no need to reference what yuzu used to call the section.

File: src/core/hle/service/ldr/ldr.cpp [hunk 2]
```diff
@@ -39,47 +39,68 @@ constexpr ResultCode ERROR_NOT_INITIALIZED{ErrorModule::Loader, 87};
...
     u32_le bss_size;
-    INSERT_PADDING_WORDS(1);
+    u32_le reserved;
     std::array<u8, 0x20> build_id;
-    INSERT_PADDING_BYTES(0x20);
+    u32_le dso_handle_offset;
+    u32_le unused_2;
+    // .apiInfo, .dynstr, .dynsym
+    std::array<SegmentHeader, 0x3> segment_headers_2;
```
Comment (lioncash) [inline src/core/hle/service/ldr/ldr.cpp:100]:
```suggestion
    std::array<SegmentHeader, 3> segment_headers_2;
```

File: src/core/hle/service/ldr/ldr.cpp [hunk 3]
```diff
@@ -39,47 +39,68 @@ constexpr ResultCode ERROR_NOT_INITIALIZED{ErrorModule::Loader, 87};
...
-    u64_le title_id;
+    u32_le certification_signature_key_generation; // 9.0.0+
+    u64_le reserved;
+    Certification certification;
+    std::array<u8, 0x100> signature;
+    u64_le application_id;
     u32_le size;
-    INSERT_PADDING_BYTES(4);
+    u8 nrr_kind; // 7.0.0+
+    std::array<u8, 0x3> reserved_2;
```
Comment (lioncash) [inline src/core/hle/service/ldr/ldr.cpp:62]:
```suggestion
    std::array<u8, 3> reserved_2;
```

File: src/core/hle/service/ldr/ldr.cpp [hunk 4]
```diff
@@ -39,42 +39,61 @@ constexpr ResultCode ERROR_NOT_INITIALIZED{ErrorModule::Loader, 87};
...
-    INSERT_PADDING_BYTES(12);
-    u64_le title_id_mask;
-    u64_le title_id_pattern;
-    INSERT_PADDING_BYTES(16);
-    std::array<u8, 0x100> modulus;
-    std::array<u8, 0x100> signature_1;
-    std::array<u8, 0x100> signature_2;
-    u64_le title_id;
+    u32_le certification_signature_key_generation; // 9.0.0+
+    u64_le reserved;
```
Comment (ogniK5377) [inline src/core/hle/service/ldr/ldr.cpp:58]:
```suggestion
    INSERT_PADDING_WORDS(2);
```

File: src/core/hle/service/ldr/ldr.cpp [hunk 5]
```diff
@@ -39,42 +39,61 @@ constexpr ResultCode ERROR_NOT_INITIALIZED{ErrorModule::Loader, 87};
...
 constexpr std::size_t MAXIMUM_MAP_RETRIES{0x200};
 
+constexpr std::size_t TEXT_INDEX{0};
+constexpr std::size_t RO_INDEX{1};
+constexpr std::size_t DATA_INDEX{2};
+
+struct NRRCertification {
+    u64_le application_id_mask;
+    u64_le application_id_pattern;
+    std::array<u8, 0x10> reserved;
```
Comment (ogniK5377) [inline src/core/hle/service/ldr/ldr.cpp:49]:
```suggestion
    INSERT_PADDING_BYTES(0x10);
```

File: src/core/hle/service/ldr/ldr.cpp [hunk 6]
```diff
@@ -39,42 +39,61 @@ constexpr ResultCode ERROR_NOT_INITIALIZED{ErrorModule::Loader, 87};
...
-    u64_le title_id;
+    u32_le certification_signature_key_generation; // 9.0.0+
+    u64_le reserved;
+    NRRCertification certification;
+    std::array<u8, 0x100> signature;
+    u64_le application_id;
     u32_le size;
-    INSERT_PADDING_BYTES(4);
+    u8 nrr_kind; // 7.0.0+
+    std::array<u8, 3> reserved_2;
```
Comment (ogniK5377) [inline src/core/hle/service/ldr/ldr.cpp:64]:
```suggestion
    INSERT_PADDING_BYTES(3);
```

File: src/core/hle/service/ldr/ldr.cpp [hunk 7]
```diff
@@ -39,42 +39,61 @@ constexpr ResultCode ERROR_NOT_INITIALIZED{ErrorModule::Loader, 87};
...
     u32_le hash_offset;
     u32_le hash_count;
-    INSERT_PADDING_BYTES(8);
+    u64_le reserved_3;
```
Comment (ogniK5377) [inline src/core/hle/service/ldr/ldr.cpp:67]:
```suggestion
    INSERT_PADDING_WORDS(2);
```

File: src/core/hle/service/ldr/ldr.cpp [hunk 8]
```diff
@@ -39,42 +39,61 @@ constexpr ResultCode ERROR_NOT_INITIALIZED{ErrorModule::Loader, 87};
...
+struct SegmentHeader {
+    u32_le memory_offset;
+    u32_le memory_size;
+};
+static_assert(sizeof(SegmentHeader) == 0x8, "SegmentHeader has invalid size.");
 
 struct NROHeader {
-    INSERT_PADDING_WORDS(1);
+    // Switchbrew calls this "Start" (0x10)
+    u32_le unused;
```
Comment (ogniK5377) [inline src/core/hle/service/ldr/ldr.cpp:79]:
```suggestion
    INSERT_PADDING_WORDS(1);
```

File: src/core/hle/service/ldr/ldr.cpp [hunk 9]
```diff
@@ -39,42 +39,61 @@ constexpr ResultCode ERROR_NOT_INITIALIZED{ErrorModule::Loader, 87};
...
     u32_le mod_offset;
-    INSERT_PADDING_WORDS(2);
+    u64_le padding;
```
Comment (ogniK5377) [inline src/core/hle/service/ldr/ldr.cpp:81]:
```suggestion
    INSERT_PADDING_WORDS(2);
```

File: src/core/hle/service/ldr/ldr.cpp [hunk 10]
```diff
@@ -39,42 +39,61 @@ constexpr ResultCode ERROR_NOT_INITIALIZED{ErrorModule::Loader, 87};
...
-    u32_le text_size;
-    u32_le ro_offset;
-    u32_le ro_size;
-    u32_le rw_offset;
-    u32_le rw_size;
+    // .text, .ro, .data
+    std::array<SegmentHeader, 3> segment_headers;
     u32_le bss_size;
-    INSERT_PADDING_WORDS(1);
+    u32_le reserved;
```
Comment (ogniK5377) [inline src/core/hle/service/ldr/ldr.cpp:91]:
```suggestion
    INSERT_PADDING_WORDS(1);
```

File: src/core/hle/service/ldr/ldr.cpp [hunk 11]
```diff
@@ -39,42 +39,61 @@ constexpr ResultCode ERROR_NOT_INITIALIZED{ErrorModule::Loader, 87};
...
     std::array<u8, 0x20> build_id;
-    INSERT_PADDING_BYTES(0x20);
+    u32_le dso_handle_offset;
+    u32_le unused_2;
```
Comment (ogniK5377) [inline src/core/hle/service/ldr/ldr.cpp:94]:
```suggestion
    INSERT_PADDING_WORDS(1);
```

**Pull request merged**